### PR TITLE
Feat/pytorch lightning

### PR DESCRIPTION
**Pull request by dennisbader**
Addresses #577.

Update of #697

- i think we could let user pass all pytorch-lightning trainer parameters as a dedicted kwarg like `pl_trainer_pararms`. This will set up a trainer with some darts specifics + the user's customization
- as dicussed with @hrzn, it could be nice if user can pass their own PL trainer to fit
- one issue is how we resume training (calling fit once). Best solution I've found so far was with mandatory checkpoint saving
- further investigation of warnings required.

### CODE REVIEW

File: darts/models/forecasting/ptl_model.py [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+        self.pred_roll_size = None
+        self.pred_batch_size = None
+
+    def _compute_loss(self, output, target):
+        return self.criterion(output, target)
+
+    def configure_optimizers(self):
+        """sets up optimizers"""
+
+        # TODO: i think we can move this to to pl.Trainer(). and could probably be simplified
```
Comment (hrzn) [inline darts/models/forecasting/ptl_model.py:150]: +1

File: darts/models/forecasting/ptl_tft_model.py [hunk 2]
```diff
@@ -0,0 +1,931 @@
...
+            return super().predict(self.output_chunk_length, *args, **kwargs)[:n]
+
+    def _produce_predict_output(self, x):
+        if self.likelihood:
+            output = self(x)
+            return self.likelihood.sample(output)
+        else:
+            return self(x).squeeze(dim=-1)
+
+    def _get_batch_prediction(self, n: int, input_batch: Tuple, roll_size: int) -> torch.Tensor:
```
Comment (hrzn) [inline darts/models/forecasting/ptl_tft_model.py:524]: Couldn't this one be made part of `PLMixedCovariatesTorchModel` ?

**Review by hrzn — COMMENTED** [review]
Looks like a great start. Maybe so far my main comment concerns the PL trainer; I feel (but might be wrong) that it would be nice to change our models' signatures, in order to put everything training related in `fit()`. So for instance we could receive either the PL trainer kwargs (or the PL trainer itself?) as arguments to `fit()`. In init we could also receive training-related arguments (or PL trainer directly), but those would only be used as default if nothing is specified to `fit()` (as we are currently doing with `epochs`).

**Comment by dennisbader** [comment]
Apart from below, PL integration seems to work well.

TODO for PTL: reloading the model does not yield identical prediction results. There are currently <10 unit tests failing related to this issue.

I see two possible causes:
 1) something might be going wrong with random seeds
 2) something is going wrong in the way we save/load
      - saving:
          - TorchForecastingModel is only saved once at model creation to .pth.tar
          - Model (PLForecastingModule) and PL Trainer checkpoints are saved through the checkpoint
              callbacks every n-epochs
      - loading a checkpoint loads the following:
          - TorchForecastingModel from .pth.tar
          - Model (PLForecastingModule) is loaded from checkpoint which AFAIK recreates model from
              hyperparemeters with saved weights, biases etc. This means that optimizers, lr-scheduler
              etc are currently re-initialized (see `configure_optimizers()`). Maybe we have to save/load
              more than in the current state??
          - PL trainer is loaded (resumes) from checkpoint AND initialized with parameters
              `TorchForecastingModel.trainer_params`. Are we doing this the correct way? How to properly
              reload a trainer?

File: darts/models/forecasting/torch_forecasting_model.py [hunk 3]
```diff
@@ -1251,110 +1208,64 @@ def load_from_checkpoint(
...
-
-        Parameters
-        ----------
-        likelihood
-            The likelihood model to be used for probabilistic forecasts.
-        """
-        super().__init__(**kwargs)
-        self.likelihood = likelihood
+    @property
+    def epochs_trained(self) -> int:
```
Comment (hrzn) [inline darts/models/forecasting/torch_forecasting_model.py:1249]: Are these 4 properties used?

File: requirements/torch.txt [hunk 4]
```diff
@@ -1,2 +1,3 @@
 torch>=1.8.0
 tensorboard>=2.4.0
+pytorch-lightning>=1.5.8
```
Comment (hrzn) [inline requirements/torch.txt:3]: If you haven't done so already, might be worth testing with older versions to see if it also works, in order to have the loosest possible requirements here.

**Review by hrzn — COMMENTED** [review]
I just have a few comments, but nothing important.
This is really great stuff!!
Let's see if we can fix the dtype issue when checkpointing, and the remaining unit tests.

**Review by hrzn — APPROVED** [review]
💯

File: darts/models/forecasting/torch_forecasting_model.py [hunk 5]
```diff
@@ -59,40 +70,58 @@
...
+                "There is no file matching prefix {} in {}".format(
+                    "best-*" if best else "last-*", checkpoint_dir
+                )
+            ),
+            logger,
+        )
+
+    file_name = max(
+        checklist, key=os.path.getctime
+    )  # latest file TODO: check case where no files match
```
[the remark this answers is not in the document]
Comment (dennisbader) [reply darts/models/forecasting/torch_forecasting_model.py:111]: yep, that was just copy pasted from old version

File: darts/models/forecasting/torch_forecasting_model.py [hunk 6]
```diff
@@ -159,164 +187,254 @@ def __init__(
...
-            self.device = self._get_best_torch_device()
-        else:
-            self.device = torch.device(torch_device_str)
+        suppress_lightning_warnings(suppress_all=not show_warnings)
 
         # We will fill these dynamically, upon first call of fit_from_dataset():
-        self.model = None
-        self.train_sample = None
-        self.output_dim = None
+        self.model: Optional[PLForecastingModule] = None
```
[the remark this answers is not in the document]
Comment (dennisbader) [reply darts/models/forecasting/torch_forecasting_model.py:231]: The `Optional` here will tell type hint that `None` is also accepted.
If you don't use Optional and set it to `None`, type hint would complain as `None` is not a `PLForecastingModule`.

I will leave it as is.

File: darts/models/forecasting/torch_forecasting_model.py [hunk 7]
```diff
@@ -159,164 +187,254 @@ def __init__(
...
 
-        self.force_reset = force_reset
-        self.save_checkpoints = save_checkpoints
-        checkpoints_folder = _get_checkpoint_folder(self.work_dir, self.model_name)
-        self.checkpoint_exists = (
-            os.path.exists(checkpoints_folder)
-            and len(glob(os.path.join(checkpoints_folder, "checkpoint_*"))) > 0
+        # save tensorboard under 'darts_logs/model_name/logs/'
+        model_logger = (
+            pl_loggers.TensorBoardLogger(save_dir=log_folder, name="", version="logs")
```
[the remark this answers is not in the document]
Comment (dennisbader) [reply darts/models/forecasting/torch_forecasting_model.py:309]: Yes we can :)

File: darts/models/forecasting/torch_forecasting_model.py [hunk 8]
```diff
@@ -159,164 +187,254 @@ def __init__(
...
-        self.optimizer = _create_from_cls_and_kwargs(self.optimizer_cls, optimizer_kws)
-
-        if self.lr_scheduler_cls is not None:
-            lr_sched_kws = {k: v for k, v in self.lr_scheduler_kwargs.items()}
-            lr_sched_kws["optimizer"] = self.optimizer
-            self.lr_scheduler = _create_from_cls_and_kwargs(
-                self.lr_scheduler_cls, lr_sched_kws
+        self.trainer_params["precision"] = precision
+
+        # we need to save the initialized TorchForecastingModel as PyTorch-Lightning only saves module checkpoints
```
[the remark this answers is not in the document]
Comment (dennisbader) [reply darts/models/forecasting/torch_forecasting_model.py:402]: I think we do as only the PyTorch-Lightning related data (`PLForecastingModule` and Trainer states) is saved through the checkpoints.

So when we load from checkpoint, first we need to load our  `TorchForecastingModel` instance and then load the model and trainer into it.

Or do you see a way around this?

File: darts/models/forecasting/torch_forecasting_model.py (hunk 3 above)
Comment (dennisbader) [reply darts/models/forecasting/torch_forecasting_model.py]: All but `loss_fn` (now removed). They are used for direct "communication" between `ForecastingModel` and `PLForecastingModule`. For example likelihood is required for `_is_probabilistic()`.

File: darts/models/forecasting/torch_forecasting_model.py [hunk 9]
```diff
@@ -159,164 +187,254 @@ def __init__(
...
 
-        # Persist optimiser and LR scheduler parameters
-        self.optimizer_cls = optimizer_cls
-        self.optimizer_kwargs = dict() if optimizer_kwargs is None else optimizer_kwargs
-        self.lr_scheduler_cls = lr_scheduler_cls
-        self.lr_scheduler_kwargs = (
-            dict() if lr_scheduler_kwargs is None else lr_scheduler_kwargs
+        # check torch device
+        raise_if_not(
+            torch_device_str in [None, "cuda:0", "cpu"],
```
[the remark this answers is not in the document]
Comment (dennisbader) [reply darts/models/forecasting/torch_forecasting_model.py:285]: PTL Trainer should handle this with paramaters `accelerator`, `gpus`, `auto_select_gpus` and `devices`.
https://pytorch-lightning.readthedocs.io/en/stable/common/trainer.html#trainer-flags

File: darts/models/forecasting/pl_forecasting_module.py [hunk 10]
```diff
@@ -0,0 +1,581 @@
...
+        if self.likelihood:
+            return self.likelihood.compute_loss(output, target)
+        else:
+            # If there's no likelihood, nr_params=1 and we need to squeeze out the
+            # last dimension of model output, for properly computing the loss.
+            return self.criterion(output.squeeze(dim=-1), target)
+
+    def configure_optimizers(self):
+        """sets up optimizers"""
+        # TODO: i think we can move this to to pl.Trainer(). and could probably be simplified
```
[the remark this answers is not in the document]
Comment (dennisbader) [reply darts/models/forecasting/pl_forecasting_module.py:238]: Actually I checked this and unfortunately I think it only works like this.

Optimizers can only be instantiated with model.parameters(). And these are only known after the model was 
created. 

So the user will not be able to pass an optimizer object / lr scheduler himself.

[the remark this answers is not in the document]
Comment (hrzn) [reply darts/models/forecasting/pl_forecasting_module.py:238]: OK

File: requirements/torch.txt (hunk 4 above)
Comment (dennisbader) [reply requirements/torch.txt]: Yes, 1.5.0 is the lowest supported from what I've found.

File: darts/models/forecasting/block_rnn_model.py [hunk 11]
```diff
@@ -178,82 +172,123 @@ def __init__(
...
+            supported kwargs.
+            With parameter ``"callbacks"`` you can add custom or PyTorch-Lightning built-in callbacks to Darts'
+            :class:`TorchForecastingModel`. Below is an example for adding EarlyStopping to the training process.
+            The model will stop training early if the validation loss `val_loss` does not improve beyond
+            specifications. For more information on callbacks, visit:
+            `PyTorch Lightning Callbacks
+            <https://pytorch-lightning.readthedocs.io/en/stable/extensions/callbacks.html>`_
+
+            .. highlight:: python
+            .. code-block:: python
```
Comment (hrzn) [inline darts/models/forecasting/block_rnn_model.py:266]: 👍  for the example!

**Review by hrzn — APPROVED** [review]
🚀
